### ISSUE TITLE
open_by_url silently failing

### ISSUE THREAD

**Issue by gnigg** [burnash/gspread#658]
I recently upgraded from gspread 0.6.2 to 3.1.0. According to the docs, `open_by_url` should raise an exception if the spreadsheet is not found; I noticed that it does not. Instead, the command runs without error and appears to get set to spreadsheet instance which has the following output.

```
(Pdb) ss = self._client.open_by_url('http://docs.google.com/spreadsheets/d/foobar')
(Pdb) print(ss)
*** gspread.exceptions.APIError: {
  "error": {
    "code": 404,
    "message": "Requested entity was not found.",
    "status": "NOT_FOUND"
  }
}
```
Here, the `ss` variable is set showing that an exception was never raised. This is causing my exception handler for exponential backoff to be completely bypassed.

### Environment Info
Operating System: Mac OSX 10.13.6 High Sierra
Python version: 3.7.1
gspread version: 3.1.0

### Steps to reproduce
1. c = gspread.authorize(self._credentials)
2. ss = c.open_by_url('http://docs.google.com/spreadsheets/d/foobar')

### Observations
* The `ss` variable is getting set showing that no exception was raised.
* The output of `ss` shows an exception but it was never raised:
```
(Pdb) print(ss)
*** gspread.exceptions.APIError: {
  "error": {
    "code": 404,
    "message": "Requested entity was not found.",
    "status": "NOT_FOUND"
  }
}
```
* Note that it would be nice if `open_by_key` raised an exception as well (it too does not)

**Comment by gnigg** [comment]
Not really sure what is happening here. If I do any operation on the `ss` object, then the exception gets raised. As a hack, all I did was:
```
c = gspread.authorize(self._credentials)
ss = c.open_by_url('http://docs.google.com/spreadsheets/d/foobar')
str(ss)
```
That last line will ultimately cause the intended exception.

**Comment by leahein** [comment]
I am experiencing the same issue for `open_by_key` on version  3.2.0.
Is there a fix for this?

**Comment by zachliu** [comment]
I just checked the source code, this thing should be easy to fix

* In `0.6.2`, this is why `SpreadsheetNotFound` can be raised: https://github.com/burnash/gspread/blob/6b1bb1ef6db8b318e3da5053ab173add853d3fe0/gspread/client.py#L106

* In `3.2.0`, just add the `try-except` block back:
https://github.com/burnash/gspread/blob/bcd95f751b49e8392dd65dde0cab2a2f0a46b497/gspread/client.py#L145

**Comment by lavigne958** [comment]
This issue needs to be addresses I think.
One can simply do a call to `fetch_sheet_metadata()` right after a call to `open(_by_url|_by_key)`.

I am thinking about doing it in the open method, this way we'll catch wrong name/url/key etc and we'll be able to populate the SpreadSheet instance properties.

**Comment by lavigne958** [comment]
This has been solved in version v5.0.0

Closing this issue.
